core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -0,0 +1,2103 @@
+                    }
+                }
+            }
+
+            outputSelectors.forEach(((planNodeId, outputSelector) -> splits.put(planNodeId, createOutputSelectorSplit(outputSelector))));
+
+            int attempt = maxTaskExecutionAttempts - partition.getRemainingAttempts();
+            ExchangeSinkInstanceHandle exchangeSinkInstanceHandle = exchange.instantiateSink(partition.getExchangeSinkHandle(), attempt);
+            SpoolingOutputBuffers outputBuffers = SpoolingOutputBuffers.createInitial(exchangeSinkInstanceHandle, sinkPartitioningScheme.getPartitionCount());
+            Optional<RemoteTask> task = stage.createTask(
+                    node,
+                    partitionId,
+                    attempt,
+                    sinkPartitioningScheme.getBucketToPartitionMap(),
+                    outputBuffers,
+                    splits,
+                    noMoreSplits,
+                    Optional.of(partition.getMemoryRequirements().getRequiredMemory()));
+            task.ifPresent(remoteTask -> partition.addTask(remoteTask, outputBuffers));
+            return task;
+        }
+
+        public void updateExchangeSinkInstanceHandle(TaskId taskId)
+        {
+            if (getState().isDone()) {
+                return;
+            }
+            StagePartition partition = getStagePartition(taskId.getPartitionId());
+            ExchangeSinkInstanceHandle exchangeSinkInstanceHandle = exchange.updateSinkInstanceHandle(partition.getExchangeSinkHandle(), taskId.getAttemptId());
+            partition.updateExchangeSinkInstanceHandle(taskId, exchangeSinkInstanceHandle);
+        }
+
+        public void taskFinished(TaskId taskId, TaskStatus taskStatus)
+        {
+            if (getState().isDone()) {
+                return;
+            }
+
+            int partitionId = taskId.getPartitionId();
+            StagePartition partition = getStagePartition(partitionId);
+            exchange.sinkFinished(partition.getExchangeSinkHandle(), taskId.getAttemptId());
+            SpoolingOutputStats.Snapshot outputStats = partition.taskFinished(taskId);
+
+            if (!remainingPartitions.remove(partitionId)) {
+                // a different task for the same partition finished before
+                return;
+            }
+
+            updateOutputSize(outputStats);
+
+            // task tescriptor has been created
+            if (partition.getOpenTaskDescriptor().isEmpty()) {
+                taskDescriptorStorage.remove(stage.getStageId(), partitionId);
+            }
+
+            partitionMemoryEstimator.registerPartitionFinished(
+                    queryStateMachine.getSession(),
+                    partition.getMemoryRequirements(),
+                    taskStatus.getPeakMemoryReservation(),
+                    true,
+                    Optional.empty());
+
+            sinkOutputSelectorBuilder.include(exchange.getId(), taskId.getPartitionId(), taskId.getAttemptId());
```

Comment:
This is not yet needed. It will be needed once we implement task splitting, when a certain partition has to be excluded from the result set. (e.g.: splitting task 1 into task 2 and task 3)

Reply:
When a task fails a different attempt is scheduled. If a task runs out of attempts we fail the query. The `exclude` semantics are designed to exclude an entire "taskPartition", currently it is not possible to give a signal to exclude certain attempt (it can be done only by "including" a different, succesfull attempt). I wonder how important it is to be able to deliver the signal about excluding a certain attempt earlier?